# ML-02 — Research Question and Provisional Lane

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

**Chosen Lane:** **Lane 2 – Refresh / Content Opportunity Scoring**

### Why I chose this lane

I chose the **Refresh / Content Opportunity Scoring** lane because it addresses a real business decision: determining which website pages should be refreshed first. Rather than simply analyzing historical performance, this lane focuses on prioritizing pages for action, making the output directly useful for SEO specialists and content teams. It also aligns well with my interest in machine learning and provides an opportunity to build a ranking system using multiple search and engagement signals.

---

## 2. The question: decision, action, cost of a wrong call

### Decision, Action, and Cost of a Wrong Recommendation:

### Decision

The goal is to help determine **which website pages should be refreshed first**. Since content teams have limited time and resources, they cannot update every page. This project aims to prioritize pages that are most likely to benefit from a content refresh.

### Who Acts on It?

The recommendations will be used by **SEO specialists** and **content editors**. They can use the ranked list to decide which pages to update, optimize, or monitor first.

### Action:

For each page, the model produces a **Refresh Priority Score**. Pages with the highest scores are reviewed and refreshed before lower-priority pages, allowing the team to focus on content with the greatest potential impact.

### Cost of a Wrong Recommendation:

**False Positive (High priority but doesn't need refreshing):**
- Wastes editorial time and effort.
- Uses resources on pages with little expected benefit.

**False Negative (Needs refreshing but not recommended):**
- Valuable pages may continue to lose search visibility.
- Organic traffic and user engagement may decline.
- Potential business opportunities may be missed.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [28]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [29]:
!git clone https://github.com/Dev-hashh/flyrank-ml-internship-starter.git

Cloning into 'flyrank-ml-internship-starter'...
remote: Enumerating objects: 115, done.
remote: Counting objects: 100% (115/115), done.
remote: Compressing objects: 100% (84/84), done.
remote: Total 115 (delta 31), reused 85 (delta 15), pack-reused 0 (from 0)
Receiving objects: 100% (115/115), 1.87 MiB | 8.27 MiB/s, done.
Resolving deltas: 100% (31/31), done.


In [30]:
import os
%cd /content/flyrank-ml-internship-starter
# Define the path to the skills/README.md file
readme_path = 'skills/README.md'

# Check if the file exists before trying to read it
if os.path.exists(readme_path):
    with open(readme_path, 'r') as f:
        readme_content = f.read()
    print(readme_content)
else:
    print(f"Error: The file '{readme_path}' was not found.")

/content/flyrank-ml-internship-starter
# Skills — the router

This folder is a small library of **skills**: focused instruction files your AI assistant loads
one at a time. One skill per task keeps the assistant sharp — its context window is small, and
filling it with everything makes it worse at the one thing you need.

**How to use it (repo-reading agents — Claude Code, Cursor, Codex):** they find this file
automatically via `AGENTS.md` / `CLAUDE.md`. Just tell your assistant which task you're doing.

**Using a chat-only assistant (ChatGPT / Gemini in a browser)?** Open the skill file on GitHub,
copy its whole content, and paste it into your chat before asking for help. That's it.

## The table — find your task, load ONE skill

| Your task | Load this skill | Also load for data work |
|---|---|---|
| Any task — how to work with your assistant at all | `directing-your-ai-assistant/SKILL.md` | — |
| Pick a lane, frame your question (ML-02, ML-03) | `framing-ml-problems/SKILL.md` | `fly

In [31]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Optional
import seaborn as sns

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

df = pd.read_csv(
    "/content/flyrank-ml-internship-starter/data/raw/content_refresh_anonymized.csv"
)

In [32]:
print("Shape: ", df.shape)


Shape:  (30000, 44)


In [33]:
print("Head:\n", df.head())

Head:
              content_id          client_id  search_volume  competition  \
0  content_304f48230142  client_f369cb89fc           10.0         0.67   
1  content_a1fb4e703a9e  client_4e07408562           90.0         0.01   
2  content_9aa793d4d895  client_7f2253d7e2            0.0         0.00   
3  content_331d6c4de07b  client_19581e27de           10.0         0.00   
4  content_d99b7a2d90ca  client_3fdba35f04            0.0         0.00   

  competition_level   cpc     content_type    main_intent  word_count  \
0              HIGH  2.05  keyword article  transactional      3221.0   
1               LOW  0.05  keyword article  informational      2481.0   
2               LOW  0.00  keyword article  informational      3515.0   
3               LOW  0.00  keyword article     commercial         NaN   
4               LOW  0.00  keyword article  informational      2803.0   

   char_count provider_used              model_used  impressions_90d  \
0     20457.0           NaN        ge

In [34]:
print("Col list: ",df.columns.tolist())

Col list:  ['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']


In [35]:
print("Info: \n", df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 44 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   content_id              30000 non-null  object 
 1   client_id               30000 non-null  object 
 2   search_volume           27532 non-null  float64
 3   competition             27532 non-null  float64
 4   competition_level       27390 non-null  object 
 5   cpc                     27532 non-null  float64
 6   content_type            30000 non-null  object 
 7   main_intent             27626 non-null  object 
 8   word_count              22301 non-null  float64
 9   char_count              22301 non-null  float64
 10  provider_used           8562 non-null   object 
 11  model_used              24267 non-null  object 
 12  impressions_90d         30000 non-null  int64  
 13  clicks_90d              30000 non-null  int64  
 14  pageviews_90d           30000 non-null

In [36]:
df.describe().T


,count,mean,std,min,25%,50%,75%,max
search_volume,27532.0,158.882391,1518.270825,0.0,0.0,10.00,20.00,74000.00
competition,27532.0,0.146954,0.285241,0.0,0.0,0.00,0.13,1.00
cpc,27532.0,0.485342,2.101560,0.0,0.0,0.00,0.00,100.36
word_count,22301.0,3107.760325,1452.382598,8.0,2413.0,2877.00,3666.00,9546.00
char_count,22301.0,20665.277835,10115.344042,40.0,15644.0,19116.00,24011.00,111158.00
impressions_90d,30000.0,5200.366300,16838.019547,1.0,81.0,731.00,3615.25,517715.00
clicks_90d,30000.0,16.097333,75.076958,0.0,0.0,1.00,7.00,4178.00
pageviews_90d,30000.0,49.942467,152.101430,0.0,2.0,8.00,33.00,5998.00
sessions_90d,30000.0,37.066633,107.069131,1.0,2.0,7.00,27.00,4345.00
users_90d,30000.0,35.937700,103.748185,1.0,2.0,7.00,27.00,4913.00


## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

### Dataset Summary

The starter dataset contains **30,000 website pages** described by **44 features**, including search performance, engagement metrics, content metadata, and freshness signals.

A few observations that support the **Refresh / Content Opportunity Scoring** lane:

- The average page receives **5,200 impressions** over 90 days, but the median is only **731**, indicating that page visibility varies widely across the dataset.
- The median **CTR is 0.07%**, while the average is **0.51%**, suggesting many pages have low click-through rates and may benefit from optimization.
- The median page age is **236 days**, with pages ranging from **90 to 564 days** old, indicating that some content is significantly older than others and may require refreshing.

These observations suggest that there is substantial variation in visibility, engagement, and content freshness, making it worthwhile to build a scoring system that prioritizes pages for content refresh.

## What My Work Can and Cannot Claim

### What this project can claim

- It identifies **observed patterns** in search performance, engagement, and content freshness.
- It provides a **decision-support ranking** that helps prioritize pages for content refresh.
- It measures associations between available signals and refresh opportunities using the provided dataset.

### What this project cannot claim

- It does **not** prove that refreshing a page will cause better search rankings or increased traffic.
- It does **not** predict Google's ranking algorithm or future search behavior.
- It does **not** guarantee that every recommended page will improve after being refreshed.

The recommendations produced by this project should be used as **decision support** rather than definitive predictions.

## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.